catalog/app/utils/spreadsheets/spreadsheets.ts

```diff
@@ -0,0 +1,168 @@
+import * as dateFns from 'date-fns'
+import * as R from 'ramda'
+import xlsx from 'xlsx'
+
+import * as jsonSchema from 'utils/json-schema'
+import pipeThru from 'utils/pipeThru'
+
+type MetadataValue = $TSFixMe
+
+type JsonSchema = $TSFixMe
+
+export function rowsToJson(rows: MetadataValue[][]) {
+  // TODO: R.pipe(R.tail, R.length, R.max(memo)) ?
+  const maxSize = rows.reduce((memo, row) => R.max(memo, R.tail(row).length), 0)
+  return pipeThru(rows)(
+    R.map(([key, ...values]) => {
+      const nullsTail = R.repeat(null, maxSize - values.length)
+      return [key, R.concat(values, nullsTail)]
+    }),
+    R.fromPairs,
+  )
+}
+
+export function parseSpreadsheet(
+  sheet: xlsx.WorkSheet,
+  transpose: boolean,
+): Record<string, MetadataValue> {
+  const rows = xlsx.utils.sheet_to_json<MetadataValue>(sheet, {
+    header: 1,
+  })
+  return rowsToJson(transpose ? R.transpose(rows) : rows)
+}
+
+export function readSpreadsheet(file: File): Promise<xlsx.WorkSheet> {
+  return new Promise((resolve, reject) => {
+    const reader = new FileReader()
+    reader.onabort = () => {
+      reject(new Error('abort'))
+    }
+    reader.onerror = () => {
+      reject(reader.error)
+    }
+    reader.onload = () => {
+      const workbook = xlsx.read(reader.result, { type: 'array', cellDates: true })
+      resolve(workbook.Sheets[workbook.SheetNames[0]])
+    }
+    reader.readAsArrayBuffer(file)
+  })
+}
+
+export function scoreObjectDiff(obj1: {}, obj2: {}): number {
+  const keys = Object.keys(obj1)
+  return keys.reduce((memo, key) => {
+    if (key in obj2) return memo + 1
+    return memo
+  }, 0)
+}
+
+function parseJSON(str: string | number | boolean) {
+  if (typeof str !== 'string') return str
+
+  try {
+    return JSON.parse(str)
+  } catch (e) {
+    return str
+  }
+}
+
+const isDate = (value: MetadataValue, schema?: JsonSchema) =>
+  value instanceof Date ||
+  (schema && schema.type === 'string' && schema.format === 'date' && value)
+
+const isArrayOfDates = (value: MetadataValue, schema?: JsonSchema) =>
+  isDate(value, schema?.items) || isDate(value, schema?.contains)
+
+const isList = (value: MetadataValue, schema?: JsonSchema) =>
+  schema && jsonSchema.isSchemaArray(schema) && typeof value === 'string'
+
+const isArrayOfLists = (value: MetadataValue, schema?: JsonSchema) =>
```

Comment:
i believe we should use List/Array consistently, like, this is more like `ArrayOfArrays`

Reply:
I can do this. My reasoning was, it's not array yet, it's a string with the list in it: `"1, 2, 3, 4, 5"`